tket/src/Converters/ChoiMixTableauConverters.cpp

```diff
@@ -30,641 +29,732 @@ ChoiMixTableau circuit_to_cm_tableau(const Circuit& circ) {
+    for (unsigned s = 0; s < r; ++s)
+      if (tab.tab_.zmat_(s, qb_col)) tab.tab_.row_mult(r, s);
+    // Initialise with correct phase
+    if (row.second.coeff == -1.)
+      out_circ_tp.add_op<Qubit>(OpType::X, {initialised_qb});
+    tab.remove_row(r);
     zero_initialised.insert(initialised_qb);
+    tab.discard_qubit(initialised_qb, ChoiMixTableau::TableauSegment::Output);
+  }
+}
+
+void ChoiMixBuilder::solve_collapsed_subspace() {
+  // Solving the initialised subspace will have preserved the upper echelon form
+  // of the collapsed subspace; reduce the inputs of the collapsed space to a
+  // minimal set of qubits using CX gates
+  unsigned n_ins = tab.get_n_inputs();
+  unsigned n_outs = tab.get_n_outputs();
+  MatrixXb subtableau = tab.tab_.zmat_.topLeftCorner(tab.get_n_rows(), n_ins);
+  std::vector<std::pair<unsigned, unsigned>> col_ops =
+      leading_column_gaussian_col_ops(subtableau);
+  for (const std::pair<unsigned, unsigned>& op : col_ops) {
+    tab.tab_.apply_CX(op.second, op.first);
+    ChoiMixTableau::col_key_t ctrl = tab.col_index_.right.at(op.second);
+    ChoiMixTableau::col_key_t trgt = tab.col_index_.right.at(op.first);
+    in_circ.add_op<Qubit>(OpType::CX, {ctrl.first, trgt.first});
+  }
+  // Since row multiplications will unsolve the inputs, we cannot get the output
+  // segment into upper echelon form for the same CX-saving trick; instead we
+  // accept just removing any qubits that are now unused after solving the
+  // initialised subspace
+  remove_unused_qubits();
+  tab.canonical_column_order(ChoiMixTableau::TableauSegment::Input);
+  // Solve the output segment using CX gates
+  n_ins = tab.get_n_inputs();
+  n_outs = tab.get_n_outputs();
+  col_ops = gaussian_elimination_col_ops(
+      tab.tab_.zmat_.topRightCorner(tab.get_n_rows(), n_outs));
+  for (const std::pair<unsigned, unsigned>& op : col_ops) {
+    tab.tab_.apply_CX(n_ins + op.second, n_ins + op.first);
+    ChoiMixTableau::col_key_t ctrl = tab.col_index_.right.at(n_ins + op.second);
+    ChoiMixTableau::col_key_t trgt = tab.col_index_.right.at(n_ins + op.first);
+    out_circ_tp.add_op<Qubit>(OpType::CX, {ctrl.first, trgt.first});
+  }
+  // Connect up and remove rows and columns
+  for (unsigned r = tab.get_n_rows(); r > 0;) {
+    --r;  // Refers to the final row
+    // Check that row r has been successfully reduced
+    ChoiMixTableau::row_tensor_t row_r = tab.get_row(r);
+    if (row_r.first.string.map.size() != 1 ||
+        row_r.first.string.map.begin()->second != Pauli::Z ||
+        row_r.second.string.map.size() != 1 ||
+        row_r.second.string.map.begin()->second != Pauli::Z)
+      throw std::logic_error(
+          "Unexpected error during collapsed subspace reduction in "
+          "ChoiMixTableau synthesis");
+    Qubit in_q = row_r.first.string.map.begin()->first;
+    Qubit out_q = row_r.second.string.map.begin()->first;
+    // Solve phase
+    if (row_r.second.coeff == -1.) {
```

Comment:
Use approximate comparison rather than `==`...
(Unless we can be certain no rounding can have occurred.)